### PR TITLE
Allow download-frozen-images to work without `go`

### PR DESCRIPTION
**Pull request by cpuguy83**
Currently we use the `go` command to read GOARCH and use the value for
matching manifests.

This change allows:

1. Specifying the arch through `TARGETARCH`
2. Falling back to `dpkg` if `go` is not available
3. Falling back to `uname -m` if `dpkg` is not available
4. A default value (amd64) if none of these commands is available.

Some of these fallbacks could definitely be better.

### CODE REVIEW

File: contrib/download-frozen-image-v2.sh [hunk 1]
```diff
@@ -201,6 +201,58 @@ handle_single_manifest_v2() {
...
+	fi
+
+	if type uname > /dev/null; then
+		uArch="$(uname -m)"
+		case "${uArch}" in
+			x86_64)
+				echo amd64
+				return 0
+				;;
+			arm|armv6l|armv7l)
```
Comment (thaJeztah) [inline contrib/download-frozen-image-v2.sh:236]: looks like shfmt is configured to want spaces here. Either update this script or we need to look what option to set on shftm to format without spaces (either way would be fine with me)

9EC68492-A6A4-42EC-8749-AAF4F02DED14

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks!!! <3

File: contrib/download-frozen-image-v2.sh [hunk 2]
```diff
@@ -201,6 +201,58 @@ handle_single_manifest_v2() {
...
 }
 
+get_target_arch() {
+	if [ -n "${TARGETARCH}" ]; then
+		echo "${TARGETARCH}"
+		return 0
+	fi
+
+	if type go > /dev/null; then
+		echo "$(go env GOARCH)"
```
Comment (tianon) [inline contrib/download-frozen-image-v2.sh:211]:
```suggestion
		go env GOARCH
```

File: contrib/download-frozen-image-v2.sh [hunk 3]
```diff
@@ -201,6 +201,58 @@ handle_single_manifest_v2() {
 	manifestJsonEntries=("${manifestJsonEntries[@]}" "$manifestJsonEntry")
 }
 
+get_target_arch() {
+	if [ -n "${TARGETARCH}" ]; then
```
Comment (tianon) [inline contrib/download-frozen-image-v2.sh:205]:
```suggestion
	if [ -n "${TARGETARCH:-}" ]; then
```

(to make sure it works if `TARGETARCH` is undefined and we `set -u`)

File: contrib/download-frozen-image-v2.sh [hunk 4]
```diff
@@ -201,6 +201,58 @@ handle_single_manifest_v2() {
...
+			armel | armhf)
+				echo "arm"
+				return 0
+				;;
+			*)
+				echo "${debArch}"
+				return 0
+				;;
+		esac
+	fi
```
Comment (tianon) [inline contrib/download-frozen-image-v2.sh:227]: This isn't exactly right -- the "little endian" architectures are dyslexic between Go and Debian (`mips64le` vs `mips64el`, etc).

I guess we could do something like this:

```suggestion
		case "${debArch}" in
			armel | armhf)
				echo "arm"
				return 0
				;;
			*64el)
				echo "${debArch%el}le"
				;;
			*)
				echo "${debArch}"
				return 0
				;;
		esac
	fi
```

File: contrib/download-frozen-image-v2.sh [hunk 5]
```diff
@@ -201,6 +201,58 @@ handle_single_manifest_v2() {
...
+		case "${uArch}" in
+			x86_64)
+				echo amd64
+				return 0
+				;;
+			arm | armv6l | armv7l)
+				echo arm
+				return 0
+				;;
+			aarch64 | armv8l)
```
Comment (tianon) [inline contrib/download-frozen-image-v2.sh:240]: If we're getting `armv8l`, we're on a 32bit chip (or using `linux32` / `setarch` to pretend), so this should probably just be `arm*` or something similar like the following in the above block instead:

```suggestion
			arm | armv[0-9]*)
				echo arm
				return 0
				;;
			aarch64)
```

File: contrib/download-frozen-image-v2.sh [hunk 6]
```diff
@@ -201,6 +201,58 @@ handle_single_manifest_v2() {
...
+				echo arm64
+				return 0
+				;;
+			*)
+				echo "${uArch}"
```
Comment (tianon) [inline contrib/download-frozen-image-v2.sh:245]: This gets a little more complicated, unfortunately.  For example, on my `mips64le` machine, `uname -m` returns just `mips64` (even though in Go, that's big-endian MIPS64 which is incompatible... :facepalm:).

That being said, this is definitely fine for `s390x` and `ppc64le` (which appear to return the expected values).

**Review by tianon — COMMENTED** [review]
I sent you a few missing bits over Slack (more uses of `go` up higher in the script, including a check for `go` that'll still make the whole script bail with missing `go`), but here's a few suggestions from my first pass :innocent: :+1:

**Comment by tianon** [comment]
Also as a side note, if you want to be more "POSIXy", I'd suggest `command -v foo > /dev/null` instead of `type foo > /dev/null` :smile:

(Doesn't really matter much here because this script is already pretty heavily Bash)

File: contrib/download-frozen-image-v2.sh (hunk 3 above)
Comment (cpuguy83) [reply contrib/download-frozen-image-v2.sh:205]: Oh nice trick!

File: contrib/download-frozen-image-v2.sh [hunk 7]
```diff
@@ -201,6 +223,68 @@ handle_single_manifest_v2() {
...
+			arm | armv[0-9]*)
+				echo arm
+				return 0
+				;;
+			aarch64)
+				echo arm64
+				return 0
+				;;
+			mips*)
+				echo >&2 "I see you are running on mips, but due to converion issues I'm not sure what arch to select"
```
Comment (cpuguy83) [inline contrib/download-frozen-image-v2.sh:271]: @wanghuaiqing2010 Tianon says you might be interested in making this work.

File: contrib/download-frozen-image-v2.sh (hunk 6 above)
Comment (cpuguy83) [reply contrib/download-frozen-image-v2.sh:276]: Fixed by above case

Comment (cpuguy83) [reply contrib/download-frozen-image-v2.sh:276]: Err... well "fixed"

File: contrib/download-frozen-image-v2.sh (hunk 7 above)
Comment (tianon) [reply contrib/download-frozen-image-v2.sh:271]: To add a little more context from my own research, my `mips64le` machine is returning `mips64` for `uname -m`, and I can't find anything in any `uname` field or even `/proc/cpuinfo` to help differentiate BE from LE. :disappointed:

**Comment by thaJeztah** [comment]
Perhaps we should also update the Dockerfile to use an alpine image as base? 

    FROM base AS frozen-images

**Comment by cpuguy83** [comment]
Should be good to go now.

@thaJeztah 🤷‍♂️ it's just an extra image that we don't currently use in the Dockerfile.

**Comment by thaJeztah** [comment]
> it's just an extra image that we don't currently use in the Dockerfile.

True as well, it was just bothering me that updating go versions also rebuilds that stage, and taking up space in my build cache if I switch between PRs/branches 😁

**Comment by cpuguy83** [comment]
Updated to use debian:buster as a base for the frozen images stage.
Not alpine since as @tianon mentioned, alpine is not available for all the arches.

**Comment by cpuguy83** [comment]
And only failures are these s390 issues we've been seeing.

**Comment by thaJeztah** [comment]
> Not alpine since as @tianon mentioned, alpine is not available for all the arches.

Oh, interesting; what arch we need is not available for alpine?

**Comment by cpuguy83** [comment]
armv5, armv8, mips

File: contrib/download-frozen-image-v2.sh [hunk 8]
```diff
@@ -201,6 +199,68 @@ handle_single_manifest_v2() {
...
```
Comment (thaJeztah) [inline contrib/download-frozen-image-v2.sh:247]:
```suggestion
				echo >&2 "I see you are running on mips, but due to conversion issues I'm not sure what arch to select"
```

(ducks)

**Comment by thaJeztah** [comment]
@tianon were all your comments addressed?

**Review by tianon — APPROVED** [review]
Yep, LGTM! :+1:

**Review by AkihiroSuda — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by thaJeztah** [comment]
s390x failure is not related; merging this one

**Pull request merged**